mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp

```diff
@@ -0,0 +1,207 @@
+//===- SYCLToGPU.cpp - SYCL to GPU Patterns -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns to convert SYCL dialect to GPU dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/SYCLToGPU/SYCLToGPU.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "mlir/Support/LLVM.h"
+#include "llvm/ADT/TypeSwitch.h"
+
```

Comment:
I have found that using debugging traces in passes is not that helpful given the already great traces provided by `-debug-only=dialect-conversion,pattern-application`, so I don't think we should take that approach to debugging this pass. Despite this, it's true my former approach was a bit too obscure and hard to debug. I have reformatted the code so that we now have two patterns: one to apply when the index argument is passed and one to apply when it is not. This way, the trace from `-debug-only=dialect-conversion,pattern-application` is explanatory enough:
```
//===-------------------------------------------===//
Legalizing operation : 'sycl.work_group_id'(0xbaf9a0) {
  %0 = "sycl.work_group_id"(%arg0) : (i32) -> index

  * Fold {
  } -> FAILURE : unable to fold

  * Pattern : 'sycl.work_group_id -> ()' {
Trying to match "(anonymous namespace)::GridOpPatternNoIndex<mlir::sycl::SYCLWorkGroupIDOp>"
"(anonymous namespace)::GridOpPatternNoIndex<mlir::sycl::SYCLWorkGroupIDOp>" result 0
  } -> FAILURE : pattern failed to match

  * Pattern : 'sycl.work_group_id -> ()' {
Trying to match "(anonymous namespace)::GridOpPatternIndex<mlir::sycl::SYCLWorkGroupIDOp>"
    ** Insert  : 'memref.alloca'(0xbe7f00)
    ** Insert  : 'arith.constant'(0xbe72e0)
    ** Insert  : 'gpu.block_id'(0xbe7340)
    ** Insert  : 'affine.store'(0xbe7390)
    ** Insert  : 'arith.constant'(0xbe7460)
    ** Insert  : 'gpu.block_id'(0xbe74c0)
    ** Insert  : 'affine.store'(0xbe7510)
    ** Insert  : 'arith.constant'(0xbe75e0)
    ** Insert  : 'gpu.block_id'(0xbe7640)
    ** Insert  : 'affine.store'(0xbe7690)
    ** Insert  : 'arith.index_cast'(0xbe1150)
    ** Insert  : 'affine.load'(0xbe7760)
    ** Replace : 'sycl.work_group_id'(0xbaf9a0)
"(anonymous namespace)::GridOpPatternIndex<mlir::sycl::SYCLWorkGroupIDOp>" result 1
```
I think this is way clearer and standard: You see the first patterns (good naming) fails to match, so the second one is applied and the operations it inserts. Also, most passes do not include debugging traces themselves.